[rocprofiler-sdk] Firmware detection in sdk along with checking fw before enabling pc-sampling

sauverma93 opened the pull request:
## Motivation

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

This PR is a re-work of #196 

This PR does the following:

Parses the firmware info for different engines and micro-controllers that are listed in sysfs path. This is done in read_topology()
Reads the info into the new API struct rocprofiler_agent_fw_info_t
Stores the rocprofiler_agent_fw_info_t inside the API struct rocprofiler_agent_v0_t
The pc-sampling service functions - rocprofiler_query_pc_sampling_agent_configurations() and rocprofiler_configure_pc_sampling_service() both functions check firmware as part of their logic
This PR contains API struct updates:

- updated rocprofiler_agent_v0_t
- added a new struct rocprofiler_agent_fw_info_t
- added a new error message enum 'ROCPROFILER_STATUS_FIRMWARE_INCOMPATIBLE' in rocprofiler_status_t

## Technical Details

<!-- Explain the changes along with any relevant GitHub links. -->

## Test Plan

<!-- Explain any relevant testing done to verify this PR. -->

## Test Result

<!-- Briefly summarize test outcomes. -->

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

sauverma93 committed:
Initial commit of firmware info parsing from sysfs & checking pc-sampling fw requirement

sauverma93 committed:
addressing review comments

sauverma93 committed:
fix static asserts and warnings

sauverma93 committed:
add rocprofiler_agent_fw_info_t to JSON load/save and updated agent.cpp agent_abi test

sauverma93 committed:
Fix pytest failures due to incorrect init of rocprofiler_agent_fw_info_t

sauverma93 committed:
address review comments

sauverma93 committed:
fix formatting issues

sauverma93 committed:
fix formatting issues - 2

sauverma93 committed:
fix warnings

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/pc_sampling/ioctl/ioctl_adapter.cpp:
```diff
@@ -586,8 +624,11 @@ ioctl_pcs_create(const rocprofiler_agent_t*       agent,
                  uint32_t*                        ioctl_pcs_id)
 {
+    auto status = check_firmware_compatibility(agent, method);
+    if(status != ROCPROFILER_STATUS_SUCCESS) return status;
```
```suggestion
   
    if(auto status = check_firmware_compatibility(agent, method);
       status != ROCPROFILER_STATUS_SUCCESS) return status;
```

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/pc_sampling/ioctl/ioctl_adapter.cpp:
```diff
@@ -586,8 +624,11 @@ ioctl_pcs_create(const rocprofiler_agent_t*       agent,
+
     pcs_ioctl_version_t pcs_ioctl_version = 0;
-    auto status = get_pcs_ioctl_version_if_kfd_supports(agent->gpu_id, &pcs_ioctl_version);
+    status = get_pcs_ioctl_version_if_kfd_supports(agent->gpu_id, &pcs_ioctl_version);
```
```suggestion
    auto status = get_pcs_ioctl_version_if_kfd_supports(agent->gpu_id, &pcs_ioctl_version);
```

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/agent.cpp:
```diff
@@ -597,6 +599,33 @@ update_agent_runtime_visibility(rocprofiler_agent_t& agent_info)
+read_fw_info(const char* fname, uint32_t drm_render_minor)
+{
+    // read firmware_info
+    std::string fw_dir =
```
```suggestion
    auto fw_dir =
```

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/agent.cpp:
```diff
@@ -597,6 +599,33 @@ update_agent_runtime_visibility(rocprofiler_agent_t& agent_info)
+    // read firmware_info
+    std::string fw_dir =
+        fmt::format("/sys/class/drm/renderD{}/device/fw_version", drm_render_minor);
+    fs::path fw_path = fs::path(fw_dir) / fname;
```
```suggestion
    auto fw_path = fs::path(fw_dir) / fname;
```

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/agent.cpp:
```diff
@@ -597,6 +599,33 @@ update_agent_runtime_visibility(rocprofiler_agent_t& agent_info)
+        return kInvalidFirmwareVersion;
+    }
+    std::ifstream fw_file(fw_path);
+    std::string   fw_value;
```
```suggestion
    auto fw_file  = std::ifstream{fw_path};
    auto fw_value = std::string{};
```

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/agent.cpp:
```diff
@@ -597,6 +599,33 @@ update_agent_runtime_visibility(rocprofiler_agent_t& agent_info)
+            return static_cast<uint32_t>(std::stoul(fw_value, nullptr, 0));
+        } catch(const std::exception&)
+        {
+            ROCP_WARNING << "Failed to parse firmware version from file: " << fw_path.string();
```
```suggestion
            ROCP_WARNING << fmt::format("Failed to parse firmware version '{}' from file '{}': {}", fw_value, fw_path.string(), e.what());
```

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/agent.cpp:
```diff
@@ -597,6 +599,33 @@ update_agent_runtime_visibility(rocprofiler_agent_t& agent_info)
+        try
+        {
+            return static_cast<uint32_t>(std::stoul(fw_value, nullptr, 0));
+        } catch(const std::exception&)
```
```suggestion
        } catch(const std::exception& e)
```

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/rocprofiler.cpp:
```diff
@@ -118,6 +118,9 @@ ROCPROFILER_STATUS_STRING(ROCPROFILER_STATUS_ERROR_AGENT_ARCH_NOT_SUPPORTED,
                           "Required permission (CAP_PERFMON) is not set, permission denied")
+ROCPROFILER_STATUS_STRING(
+    ROCPROFILER_STATUS_INCOMPATIBLE_FIRMWARE,
+    "The selected service is incompatible with the installed firmware version")
```
```suggestion
    "The requested service is incompatible with the installed firmware version")
```

jrmadsen commented on projects/rocprofiler-sdk/source/include/rocprofiler-sdk/agent.h:
```diff
@@ -117,6 +117,37 @@ ROCPROFILER_CXX_CODE(
+ * @brief Typedef for the current ::rocprofiler_agent_fw_info_v0_t
+ *
+ */
+typedef rocprofiler_agent_fw_info_v0_t rocprofiler_agent_fw_info_t;
```
```suggestion
typedef struct rocprofiler_agent_firmware_info_v0_t
{
    uint32_t mec2_version;      ///< Firmware version for MEC2 engine
    uint32_t mec_version;       ///< Firmware version for MEC engine
    uint32_t rlc_version;       ///< Firmware version for RLC engine
    uint32_t rlc_srlc_version;  ///< Firmware version for RLC SRLC engine
    uint32_t rlc_srlg_version;  ///< Firmware version for RLC SRLG engine
    uint32_t rlc_srls_version;  ///< Firmware version for RLC SRLS engine
    uint32_t sdma2_version;     ///< Firmware version for SDMA2 engine
    uint32_t sdma_version;      ///< Firmware version for SDMA engine
    uint32_t smc_version;       ///< Firmware version for SMC engine
    uint32_t sos_version;       ///< Firmware version for Secure OS (PSP)
    uint32_t ta_ras_version;    ///< Firmware version for TA RAS microcontroller
    uint32_t ta_xgmi_version;   ///< Firmware version for TA XGMI microcontroller
    uint32_t vcn_version;       ///< Firmware version for VCN engine
} rocprofiler_agent_firmware_info_v0_t;
```
No need for `size` since this will not be expanded in future. Get rid of typedef since this type will always be directly used.

jrmadsen replied:
You should document that a value of `0` means no version info was found. Also, recommend changing `*_fw_version` to `*_version` since the struct name makes it clear all these versions are related to the firmware... it is just redundant

sauverma93 replied:
sure, addressed all the above and previous suggestions. using firmware version as 0xFFFFFFFF as the invalid firmware

sauverma93 replied:
correction: invalid firmware version is set to 0
#define ROCPROFILER_FIRMWARE_VERSION_NONE 0

the doc string for rocprofiler_agent_firmware_info_v0_t has also been updated to clearly mention this detail

jrmadsen commented on projects/rocprofiler-sdk/source/include/rocprofiler-sdk/agent.h:
```diff
@@ -210,6 +241,8 @@ typedef struct rocprofiler_agent_v0_t
     int32_t                               logical_node_type_id;
     rocprofiler_agent_runtime_visiblity_t runtime_visibility;
     rocprofiler_uuid_t                    uuid;  ///< GPU only. Universally unique identifier.
+    rocprofiler_agent_fw_info_t firmware_info;  ///< GPU only. Detailed firmware version information
```
```suggestion
    rocprofiler_agent_firmware_info_v0_t fw_info;  ///< GPU only. Detailed firmware version information
```
In the future, if we need to add more firmware info, we will create `rocprofiler_agent_fw_info_v1_t` and add a field `rocprofiler_agent_fw_info_v1_t firmware_info_v1;`. Also, sure `firmware` in struct name and abbreviation (`fw`) in variable name.

MythreyaK commented on projects/rocprofiler-sdk/source/include/rocprofiler-sdk/agent.h:
```diff
@@ -210,6 +241,8 @@ typedef struct rocprofiler_agent_v0_t
     int32_t                               logical_node_type_id;
     rocprofiler_agent_runtime_visiblity_t runtime_visibility;
     rocprofiler_uuid_t                    uuid;  ///< GPU only. Universally unique identifier.
+    rocprofiler_agent_fw_info_t firmware_info;  ///< GPU only. Detailed firmware version information
```
This has to be a pointer for `rocprofiler_agent_v0_t` to be backward compatible. 

```cpp
typedef struct rocprofiler_agent_v0_t {
    ...
    rocprofiler_uuid_t                    uuid;
    rocprofiler_agent_fw_info_t firmware_info; 
    // offsets of all members below changes if rocprofiler_agent_fw_info_t has new members
    some_future_member_t        future_member; 
    ...
};
```

Or, we could create a new struct `v1` but then size is not needed in the structure.

sauverma93 replied:
Jonathan [suggested](https://github.com/ROCm/rocm-systems/pull/1159#discussion_r2392158666) that in future, we would create a v1 struct if we need to add more members.

jrmadsen commented on projects/rocprofiler-sdk/source/include/rocprofiler-sdk/cxx/enum_string.hpp:
```diff
@@ -1420,7 +1420,7 @@ ROCPROFILER_ENUM_LABEL(ROCPROFILER_STATUS_ERROR_NOT_AVAILABLE);
 ROCPROFILER_ENUM_LABEL(ROCPROFILER_STATUS_ERROR_AGENT_ARCH_NOT_SUPPORTED);
 ROCPROFILER_ENUM_LABEL(ROCPROFILER_STATUS_ERROR_PERMISSION_DENIED);
-static_assert(ROCPROFILER_STATUS_LAST == 41);
+static_assert(ROCPROFILER_STATUS_LAST == 42);
```
```suggestion
ROCPROFILER_ENUM_LABEL(ROCPROFILER_STATUS_INCOMPATIBLE_FIRMWARE);
static_assert(ROCPROFILER_STATUS_LAST == 42);
```

MythreyaK replied:
```suggestion
ROCPROFILER_ENUM_LABEL(ROCPROFILER_STATUS_INCOMPATIBLE_FIRMWARE);
static_assert(ROCPROFILER_STATUS_LAST == 42);
```

MythreyaK commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/agent.cpp:
```diff
@@ -64,6 +64,8 @@ namespace
 {
 namespace fs = ::rocprofiler::common::filesystem;
 
+constexpr uint32_t kInvalidFirmwareVersion = 0xDEADBEEF;
```
Is this a possible magic number returned in the sysfs directory?

If just being used as an invalid number, is uint32_t max preferable instead?

sauverma93 replied:
sure, changed this to 0XFFFFFFFF as it is more appropriate

sauverma93 replied:
correction: invalid firmware version is now expected as 0. 
And we no longer use an extra constexpr called kInvalidFirmwareVersion in agent.cpp

MythreyaK commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/pc_sampling/ioctl/ioctl_adapter.cpp:
```diff
@@ -48,6 +48,9 @@ namespace
 #define PC_SAMPLING_IOCTL_COMPUTE_VERSION(major, minor) ROCPROFILER_COMPUTE_VERSION(major, minor, 0)
 
+#define MINIMUM_PC_SAMPLING_MEC_FW_VERSION uint32_t(0x0000001a)
+#define MINIMUM_PC_SAMPLING_SOS_FW_VERSION uint32_t(0x00360259)
```
Nit: Could these be `constexpr` constants instead?

MythreyaK commented:
I didn't notice Jonathan already left a review, my bad!

vlaindic commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/pc_sampling/ioctl/ioctl_adapter.cpp:
```diff
@@ -48,6 +48,9 @@ namespace
 
 #define PC_SAMPLING_IOCTL_COMPUTE_VERSION(major, minor) ROCPROFILER_COMPUTE_VERSION(major, minor, 0)
 
+#define MINIMUM_PC_SAMPLING_MEC_FW_VERSION uint32_t(0x0000001a)
```
We received a complain recently that our documentation might have stating the outdate FW. Here is the issue: #1140. After sync up with Jay, I think the minimal version should be 185 (b9). Please make the doc update to reflect this as well. Also, please state that this is true for both MI300A and MI300X, as I think the #1140 is about the MI300A. 

```suggestion
#define MINIMUM_PC_SAMPLING_MEC_FW_VERSION uint32_t(0x000000b9)
```

sauverma93 replied:
Hi Vladimir, bumped the version number to the correct version - this time it mentions both MI300A and MI300X for stochastic pc-sampling

vlaindic reviewed (changes requested):
FW version update requested in [#1140](https://github.com/ROCm/rocm-systems/issues/1140)

sauverma93 committed:
Addressing review comments

sauverma93 commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/agent.cpp:
```diff
@@ -64,6 +64,26 @@ namespace
 {
 namespace fs = ::rocprofiler::common::filesystem;
 
+constexpr uint32_t kInvalidFirmwareVersion = 0xFFFFFFFF;
```
Where should I keep this constant "kInvalidFirmwareVersion" ?
We also have "INVALID_PC_SAMPLING_FIRMWARE_VERSION " in ioctl_adapter.cpp

Need suggestions for the right place, so we can use one variable, instead of having two and introducing bugs

vlaindic replied:
How about in our public API? Please check:

- https://github.com/ROCm/rocm-systems/blob/d3487970176ae89598103be193c66ba207fc9fdd/projects/rocprofiler-sdk/source/include/rocprofiler-sdk/callback_tracing.h#L198
- https://github.com/ROCm/rocm-systems/blob/d3487970176ae89598103be193c66ba207fc9fdd/projects/rocprofiler-sdk/source/include/rocprofiler-sdk/fwd.h#L657

sauverma93 committed:
Correct documentation for MI300A and MI300X MEC firmware

sauverma93 committed:
update variable names

vlaindic commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/hsa/queue_controller.cpp:
```diff
@@ -142,7 +142,8 @@ constexpr rocprofiler_agent_t default_agent =
                         .runtime_visibility         = {0, 0, 0, 0, 0},
-                        .uuid = static_cast<rocprofiler_uuid_t>(agent::uuid_view_t{})};
+                        .uuid          = static_cast<rocprofiler_uuid_t>(agent::uuid_view_t{}),
+                        .firmware_info = {}};
```
Is the INVALID value default or zero?

sauverma93 replied:
thanks for the catch! updated to have proper initialization here

sauverma93 replied:
the default invalid value for firmware version is now set as 0 in fwd.h & agent.cpp

vlaindic commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/pc_sampling/ioctl/ioctl_adapter.cpp:
```diff
@@ -525,7 +580,9 @@ ioctl_query_pcs_configs(const rocprofiler_agent_t* agent, rocp_pcs_cfgs_vec_t& r
-        rocp_configs.emplace_back(rocp_cfg);
+        if(check_firmware_compatibility(agent, get_rocp_pcs_method_from_kfd(ioctl_cfg.method)) ==
+           ROCPROFILER_STATUS_SUCCESS)
+            rocp_configs.emplace_back(rocp_cfg);
```
Maybe just a preference in case we decided to introduce some more checks in the future. Then we could avoid further nesting the code.
```suggestion
        if(check_firmware_compatibility(agent, get_rocp_pcs_method_from_kfd(ioctl_cfg.method)) !=
           ROCPROFILER_STATUS_SUCCESS)
           continue;
        rocp_configs.emplace_back(rocp_cfg);
```

vlaindic commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/pc_sampling/ioctl/ioctl_adapter.cpp:
```diff
@@ -464,6 +468,57 @@ convert_ioctl_pcs_config_to_rocp(const rocprofiler_ioctl_pc_sampling_info_t& ioc
+        }
+    }
+
+    if(method == ROCPROFILER_PC_SAMPLING_METHOD_HOST_TRAP)
+    {
+        if(agent->firmware_info.sos_version == INVALID_PC_SAMPLING_FIRMWARE_VERSION)
+        {
+            ROCP_WARNING << "Cannot determine SOS firmware version for agent-" << agent->node_id
+                         << "\nHost-Trap PC sampling may not work correctly.\n";
+            return ROCPROFILER_STATUS_ERROR_INCOMPATIBLE_FIRMWARE;
+        }
+
+        if(agent->firmware_info.sos_version < MINIMUM_PC_SAMPLING_MI300_SOS_FW_VERSION)
+        {
+            ROCP_WARNING << "Host-Trap PC sampling is not supported on agent-" << agent->node_id
+                         << " due to a firmware version mismatch\n"
+                         << "Minimum required SOS firmware version is "
+                         << MINIMUM_PC_SAMPLING_MI300_SOS_FW_VERSION << ", but found "
+                         << agent->firmware_info.sos_version << "\n";
+            return ROCPROFILER_STATUS_ERROR_INCOMPATIBLE_FIRMWARE;
+        }
+    }
```
I think this will consider both MI300A and MI300X. Thus, I think we should update the documentation accordingly. Please check the comment above.

sauverma93 replied:
both MI300A and MI300X check for the same firmware version, which is minimum required version for both

vlaindic commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/pc_sampling/ioctl/ioctl_adapter.cpp:
```diff
@@ -586,6 +643,10 @@ ioctl_pcs_create(const rocprofiler_agent_t*       agent,
+    if(auto status = check_firmware_compatibility(agent, method);
+       status != ROCPROFILER_STATUS_SUCCESS)
+        return status;
+
```
Just a personal preference (nit pickyin). How about placing this as the last check, as we:

- first check the closest level to us, which is KFD
- then if 1.16 KFD is there, we could if IOCTL supports PC sampling
- finally, we check the firmware

sauverma93 replied:
I agree, but leaving it out for now, and keeping things the same as the benefit is that if we already know the lowest level check is going to fail, we don't need to do the middle checks (various ioctl calls)

vlaindic commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/pc_sampling/ioctl/ioctl_adapter.cpp:
```diff
@@ -611,7 +672,7 @@ ioctl_pcs_create(const rocprofiler_agent_t*       agent,
     args.op              = KFD_IOCTL_PCS_OP_CREATE;
     args.gpu_id          = agent->gpu_id;
-    args.sample_info_ptr = (uint64_t)(&ioctl_cfg);
+    args.sample_info_ptr = (uint64_t) (&ioctl_cfg);
```
It's weird that the formatter didn't fix this. :D

vlaindic reviewed (changes requested):
Minor nit-picking proposals. I'm fine even without them. Thanks!

vlaindic commented on projects/rocprofiler-sdk/source/docs/how-to/using-pc-sampling.rst:
```diff
@@ -57,7 +57,7 @@ The preceding output shows that the GPU supports PC sampling with the ``ROCPROFI
    To ensure that you have the latest fixes, check if you have the correct firmware versions installed:
 
    For host-trap PC-sampling on MI300X: PSP TOS Firmware >= version 00.36.02.59 or 0x00360259
```
I think we might have to add MI300A, because in the ioctl_adapter.cpp we do not make any distinction between MI300A and MI300X.

sauverma93 replied:
updated doc to mention this is for both MI300A and MI300X

sauverma93 committed:
address review comments

sauverma93 committed:
fix incorrect initialization of firmware_info in queue_controller.cpp

sauverma93 committed:
minor formatting issue fix

jrmadsen commented on projects/rocprofiler-sdk/source/include/rocprofiler-sdk/fwd.h:
```diff
@@ -130,6 +132,11 @@ typedef enum rocprofiler_agent_type_t  // NOLINT(performance-enum-size)
+/**
+ * @brief The NULL value of a firmware version.
+ */
+#define ROCPROFILER_FIRMWARE_VERSION_NONE UINT32_MAX
```
We usually use zero for "none". And using this value makes checks like this error prone:

```cpp
if(agent_v.firmware_info.mec2_version < minimum_mec2_version)
{
    ROCP_FATAL << fmt::format("MEC2 firmware version is {}. Minimum required: {}", agent_v.firmware_info.mec2_version, minimum_mec2_version);
}
```

Since `UINT32_MAX` will always be >= whatever the minimum version is.

sauverma93 replied:
I've set the invalid version to 0 now
#define ROCPROFILER_FIRMWARE_VERSION_NONE 0

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/pc_sampling/ioctl/ioctl_adapter.cpp:
```diff
@@ -464,6 +468,57 @@ convert_ioctl_pcs_config_to_rocp(const rocprofiler_ioctl_pc_sampling_info_t& ioc
+            return ROCPROFILER_STATUS_ERROR_INCOMPATIBLE_FIRMWARE;
+        }
+
+        if(agent->firmware_info.mec_version < MINIMUM_PC_SAMPLING_MI300_MEC_FW_VERSION)
```
Exactly my point in previous comment. If `.mec_version = ROCPROFILER_FIRMWARE_VERSION_NONE`, then this is always true. It will be very easy down the road for someone to forget to check `.mec_version == INVALID_PC_SAMPLING_FIRMWARE_VERSION`

vlaindic replied:
I believe this comment is resolved, correct? @sauverma93

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/tests/agent.cpp:
```diff
@@ -109,9 +109,10 @@ TEST(rocprofiler_lib, agent_abi)
     // Add test for offset of new field above this. Do NOT change any existing values!
 
-    constexpr auto expected_rocp_agent_size = 312;
+    constexpr auto expected_rocp_agent_size = 376;
```
Shouldn't you add some unit tests for `.firmware_info`???

sauverma93 replied:
added a test called "agent_fw_info" that compares the populated fw_info directly against sysfs values

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/agent.cpp:
```diff
@@ -606,6 +626,35 @@ update_agent_runtime_visibility(rocprofiler_agent_t& agent_info)
+    }
+    auto fw_file  = std::ifstream{fw_path};
+    auto fw_value = std::string{};
+    if(fw_file && std::getline(fw_file, fw_value) && !fw_value.empty())
```
- You need to log if `fw_path` cannot be read. 
- You need to log if the line is empty. 
- You should use `rocprofiler::sdk::parse::tokenize` from `<rocprofiler-sdk/cxx/details/tokenize.hpp>` on the line and check to make sure there aren't unexpected entries. 
- And I'd recommend using `rocprofiler::sdk::parse::from_string<uint32_t>(...)` here.

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/agent.cpp:
```diff
@@ -606,6 +626,35 @@ update_agent_runtime_visibility(rocprofiler_agent_t& agent_info)
+{
+    // read firmware_info
+    auto fw_dir  = fmt::format("/sys/class/drm/renderD{}/device/fw_version", drm_render_minor);
+    auto fw_path = fs::path(fw_dir) / fname;
```
```suggestion
    auto fw_path  = fmt::format("/sys/class/drm/renderD{}/device/fw_version/{}", drm_render_minor, fname);
```

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/agent.cpp:
```diff
@@ -606,6 +626,35 @@ update_agent_runtime_visibility(rocprofiler_agent_t& agent_info)
 using unique_agent_t = std::unique_ptr<rocprofiler_agent_t, void (*)(rocprofiler_agent_t*)>;
 
+uint32_t
+read_fw_info(const char* fname, uint32_t drm_render_minor)
```
```suggestion
read_fw_info(std::string_view fname, uint32_t drm_render_minor)
```

sauverma93 committed:
Address PC sampling firmware review comments

- Use 0 for ROCPROFILER_FIRMWARE_VERSION_NONE
- Improve read_fw_info logging, tokenize, and sysfs path
- Add firmware_info ABI and agent tests
- Run firmware check after KFD/ioctl checks
- Document rocm-smi and debugfs firmware verification

vlaindic commented on projects/rocprofiler-sdk/source/docs/how-to/using-pc-sampling.rst:
```diff
@@ -53,21 +53,69 @@ The output lists if ``rocprofv3`` supports PC sampling on the GPU and the suppor
+------------------------------------------------------
+
+Stochastic PC sampling on MI300 is also associated with **MEC feature version 50** in release notes.
+The SDK does **not** read feature version today—only the firmware revision fields above.
```
What exactly is the `feature version`?

sauverma93 replied:
A firmware is identified by two numbers: 
- firmware version
- feature version

feature version changes less often. Here is what I grabbed from AI:

"
Feature version and firmware version are two different identifiers for the same firmware ucode

feature version comes from the ucode_feature_version field in the AMDGPU firmware binary header (embedded in the .bin file when the driver loads MEC). Think of it as:

Feature version → “what generation of ucode interface is this?” (compatibility / capability family)
Firmware version → “which exact build within that family is running?” (bugfix / release iteration)
The kernel driver uses both for compatibility. For example, in gfx_v9_0.c it may require things like mec_feature_version >= 46 and mec_fw_version >= 0x... before enabling certain CP behaviors.

Feature version changes less often and usually bumps when the ucode ABI or supported hardware features change. Firmware version changes with each firmware release/bugfix."

sauverma93 replied:
we need to read sysfs with "sudo" in order to get both firmware version and feature version. 

```
sudo cat /sys/kernel/debug/dri/1/amdgpu_firmware_info | grep MEC
MEC feature version: 50, firmware version: 0x000000c1
```

In SDK, we don't want to force users to use sudo just to start pc-sampling. So we chose another path -

https://github.com/ROCm/rocm-systems/pull/1159/changes#r3544720040

vlaindic replied:
And for us to know what FW is whitelisted, firmware version is sufficient, correct?

sauverma93 replied:
After checking with Shweta and Jay, got to know the following:

1. The "firmware version" and "feature version" are independent. In theory the driver should only depend on feature versions and not firmware versions, but the feature version is not always updated consistently enough to reflect driver-dependent changes.
2. The firmware version doesn't go backwards for a given ASIC. It starts over at zero for new IP versions.

In conclusion - we are fine with checking only firmware version (which we are already doing in this PR) and we should always accompany firmware version check with the agent check

vlaindic replied:
I agree, as long as we first check the ASIC (which we do now), then firmware version should be fine. We might think about documenting this in the comments in our code.

vlaindic commented on projects/rocprofiler-sdk/source/docs/how-to/using-pc-sampling.rst:
```diff
@@ -53,21 +53,69 @@ The output lists if ``rocprofv3`` supports PC sampling on the GPU and the suppor
+
+To view both feature and firmware version for every engine, use the AMDGPU debugfs interface. This
+requires ``sudo`` and the correct DRI index under ``/sys/kernel/debug/dri/``:
+
+.. code-block:: bash
+
+  # List available debugfs DRI indices (one per DRM card)
+  ls /sys/kernel/debug/dri/
+
+  # Replace <N> with the index that corresponds to your GPU (often matches the card number)
+  sudo grep -E "MEC|SOS" /sys/kernel/debug/dri/<N>/amdgpu_firmware_info
+
+Example debugfs lines:
+
+.. code-block:: text
+
+  MEC feature version: 50, firmware version: 0x000000c0
+  SOS feature version: 3539549, firmware version: 0x0036025d
 
-  # To check MEC Firmware:
-  sudo cat /sys/kernel/debug/dri/1/amdgpu_firmware_info | grep MEC
+The DRI index is not always the same as the ``GPU[i]`` index shown by ``rocm-smi``. On multi-GPU
+systems, inspect each ``/sys/kernel/debug/dri/<N>/amdgpu_firmware_info`` file (or correlate
+``/sys/class/drm/card<N>/`` with the target device) until the output matches the GPU you intend to
+profile.
```
I didn't quite get why we need this if rocm-smi returns almost the same info?

sauverma93 replied:
rocm-smi shows output like so:

$ rocm-smi --showfwinfo


============================ ROCm System Management Interface ============================
================================== Firmware Information ==================================
GPU[0]          : MEC firmware version:         193
GPU[0]          : MEC2 firmware version:        193
GPU[0]          : RLC firmware version:         70
GPU[0]          : RLC SRLC firmware version:    4
GPU[0]          : RLC SRLG firmware version:    4
GPU[0]          : RLC SRLS firmware version:    4
GPU[0]          : SDMA firmware version:        27
GPU[0]          : SDMA2 firmware version:       27
GPU[0]          : SMC firmware version:         07.85.11.01
GPU[0]          : SOS firmware version:         0x0036025e
GPU[0]          : TA RAS firmware version:      27.54.02.27
GPU[0]          : TA XGMI firmware version:     32.00.00.20
GPU[0]          : VCN firmware version:         0x09118020
GPU[1]          : MEC firmware version:         193
GPU[1]          : MEC2 firmware version:        193

sauverma93 replied:
In the doc, we provided ways to check both firmware version and feature version via this command that needs sudo-

sudo cat /sys/kernel/debug/dri/1/amdgpu_firmware_info | grep MEC

But SDK can't access this path as we don't want to use sudo. So we use this way of reading firmware info:
https://github.com/ROCm/rocm-systems/pull/1159/changes#r3544720040

But this approach makes us miss feature version for the firmware. And we can't really do something about it, unless the driver itself surfaces the information in sysfs. 

So for now, we are documenting so that users know to check both firmware version and feature version

vlaindic replied:
To confirm, only firmware version is sufficient for whitelisting on our end, correct?

sauverma93 replied:
I think it might be ok. Might need to check with firmware team about their advice on how the firmware version/feature version is incremented. If the version number strictly keeps increasing, then checking only the version number should be correct

sauverma93 replied:
Posted some info in this comment:
https://github.com/ROCm/rocm-systems/pull/1159/changes#r3603876613

vlaindic commented on projects/rocprofiler-sdk/source/include/rocprofiler-sdk/agent.h:
```diff
@@ -210,6 +234,9 @@ typedef struct rocprofiler_agent_v0_t
     rocprofiler_uuid_t                    uuid;  ///< GPU only. Universally unique identifier.
+    rocprofiler_agent_firmware_info_v0_t
+        firmware_info;  ///< GPU only. Detailed firmware version information
+                        ///< for the agent's microcontrollers and engines.
```
Just to double confirm, this is not going to break the backward ABI compatibility as the `size` can be used to determine whether the field is available or not, correct?

sauverma93 replied:
I have upadted the ABI unit test but I am not sure about the call that we need to take here

vlaindic replied:
I believe we should be fine, but wouldn't have confirming this with @jrmadsen .

sauverma93 replied:
@jrmadsen will this change be backwards compatible? rocprofiler_agent_firmware_info_v0_t doesn't have a size parameter

vlaindic reviewed (commented):
Thanks for the great work. It looks good to time so far, but I added a few questions to confirm before approving. Thanks!

vlaindic reviewed (commented):
## PR Review — Firmware detection for PC sampling

Reviewing PR #1159. Full pass across generic, hygiene, and clarity rules.

---

### 🔴 CRITICAL (2)

---

#### 1. `load()` for `rocprofiler_agent_firmware_info_v0_t` takes by value — deserialized data is silently discarded

**File:** `source/include/rocprofiler-sdk/cxx/serialization/load.hpp:328`

```cpp
// BUG: takes by value — mutations inside don't reach caller
load(ArchiveT& ar, rocprofiler_agent_firmware_info_v0_t data)
```

Every other `load()` overload in this file takes by `&` reference. This one takes by value, so all the `ROCP_SDK_LOAD_DATA_FIELD` mutations are applied to a local copy and discarded. Deserialization of `firmware_info` silently produces zeros in the caller.

**Fix:** Change the signature to `rocprofiler_agent_firmware_info_v0_t& data`.

---

#### 2. Embedding `rocprofiler_agent_firmware_info_v0_t` by value inside `rocprofiler_agent_v0_t` is an ABI break — flagged by MythreyaK, not resolved

**File:** `source/include/rocprofiler-sdk/agent.h:237`

MythreyaK correctly called this out: embedding a 52-byte struct inline means any future field added to `rocprofiler_agent_firmware_info_v0_t` changes the size and offsets of `rocprofiler_agent_v0_t`, which is a public ABI-breaking change. This is structurally the same problem as why the existing `rocprofiler_agent_runtime_visiblity_t` carries a `static_assert` capping its size at `sizeof(uint32_t)`.

jrmadsen's counter-suggestion (create `_v1_t` when needed) does not resolve the issue because `sizeof(rocprofiler_agent_v0_t)` would still change.

The two compliant options are:
- **Store a pointer** (`rocprofiler_agent_firmware_info_v0_t* fw_info;`) — accessing info requires a non-null check but the agent struct stays fixed.
- **Add a `static_assert` capping `sizeof(rocprofiler_agent_firmware_info_v0_t)`** — then name the limit in the ABI test, and commit to never growing the inner struct (create `_v1_t` for extensions). This is the pattern already used for `runtime_visiblity_t`.

The ABI unit test (`tests/agent.cpp`) was updated and passes — but that's because it was updated to match the new (larger) size. That's exactly the ABI break: the test is now documenting the breakage rather than catching it.

**Fix:** Choose one of the two options before merge. Pointer is safer if the struct needs to grow; `static_assert` is simpler if 13 fields are the final set.

---

### ⚠️ IMPORTANT (5)

---

#### 3. `ROCPROFILER_FIRMWARE_VERSION_NONE = 0` — sauverma replied the invalid sentinel would change to `0xFFFFFFFF`; it was not updated

`fwd.h:138` defines `ROCPROFILER_FIRMWARE_VERSION_NONE = 0`. sauverma replied in a review thread that the invalid value would be changed to `0xFFFFFFFF`. The code was NOT updated — both `agent.cpp` and `ioctl_adapter.cpp` still alias this as their invalid sentinel.

This is a latent correctness issue: if any microcontroller reports firmware version `0` from sysfs (possible for absent or unprogrammed engines), the SDK silently blocks PC sampling with a misleading "incompatible firmware" error rather than correctly distinguishing "unreadable" from "version is actually zero."

**Fix:** Either accept `0` means "unknown/unreadable" and document it clearly in the macro comment in `fwd.h`, or change to `UINT32_MAX` (`0xFFFFFFFF`) as discussed and update all comparison logic accordingly. Either way, resolve the inconsistency between the review discussion and the committed code.

---

#### 4. `check_firmware_compatibility` is not in the anonymous namespace despite being a file-private helper

**File:** `source/lib/rocprofiler-sdk/pc_sampling/ioctl/ioctl_adapter.cpp:471`

The anonymous namespace closes at line 469. `check_firmware_compatibility` is defined between the anonymous namespace and the closing `} // namespace ioctl` — it has external linkage within `rocprofiler::pc_sampling::ioctl`. It is not declared in any header. The convention throughout this file is that all file-private helpers live in the anonymous namespace.

**Fix:** Move `check_firmware_compatibility` inside the anonymous `namespace { }` block, or mark it `static`.

---

#### 5. `std::string(agent->name) != "gfx942"` — heap-allocates and uses exact match; inconsistent with the `find("gfx94")` pattern used three screens above in the same file

**File:** `source/lib/rocprofiler-sdk/pc_sampling/ioctl/ioctl_adapter.cpp:478`

`is_pc_sampling_method_supported` (same file) uses `std::string_view agent_name = agent->name` and `agent_name.find("gfx94") == 0`. `check_firmware_compatibility` hardcodes `std::string(agent->name) != "gfx942"` — inconsistent, heap-allocates unnecessarily, and fragile if MI300A/X sub-variants ever diverge.

**Fix:**
```cpp
const std::string_view agent_name = agent->name;
if(agent_name != "gfx942") return ROCPROFILER_STATUS_SUCCESS;
```
Or use the `find("gfx94") == 0` pattern if future gfx94x variants should also be checked.

---

#### 6. Enum name mismatch: both jrmadsen and MythreyaK suggested dropping `_ERROR_` from `ROCPROFILER_STATUS_ERROR_INCOMPATIBLE_FIRMWARE` — both threads unresolved

**Files:** `source/include/rocprofiler-sdk/fwd.h:106`, `source/include/rocprofiler-sdk/cxx/enum_string.hpp:1585`

Both reviewers posted the same suggestion in separate threads; neither was resolved. Since "incompatible firmware" is an error, the `ERROR_` prefix is defensible — but leaving two unresolved suggestion threads on a public API name blocks merge.

**Fix:** Accept the reviewer suggestion (`ROCPROFILER_STATUS_INCOMPATIBLE_FIRMWARE`) or reply explaining why `ROCPROFILER_STATUS_ERROR_INCOMPATIBLE_FIRMWARE` is kept. Resolve both threads explicitly.

---

#### 7. Two identical "invalid firmware" constants in different files with no shared definition — sauverma's own open question, unresolved

- `agent.cpp:67`: `constexpr uint32_t kInvalidFirmwareVersion = ROCPROFILER_FIRMWARE_VERSION_NONE;`
- `ioctl_adapter.cpp:51`: `constexpr uint32_t INVALID_PC_SAMPLING_FIRMWARE_VERSION = ROCPROFILER_FIRMWARE_VERSION_NONE;`

Both alias the same macro. sauverma asked in a review thread where to put the shared constant and received no answer.

**Fix:** Remove both local aliases and reference `ROCPROFILER_FIRMWARE_VERSION_NONE` directly (it is already public in `fwd.h`), or define a single internal constant in a shared internal header.

---

### 💡 SUGGESTIONS (4)

---

#### 8. Copyright year on substantially modified files reads 2025 — should include 2026

Files with significant additions in this PR that still have a 2025 end year in their copyright header:
- `source/lib/rocprofiler-sdk/agent.cpp` (+99 lines)
- `source/lib/rocprofiler-sdk/pc_sampling/ioctl/ioctl_adapter.cpp` (+62 lines)
- `source/lib/rocprofiler-sdk/tests/agent.cpp` (+44 lines)
- `source/include/rocprofiler-sdk/agent.h` (+27 lines)

**Fix:** For each, change `2023-2025` → `2023-2026` (or `2024-2025` → `2024-2026`).

---

#### 9. jrmadsen's six style suggestion threads are not yet applied or rejected

All six are open GitHub suggestion blocks:
- Use `fs::path` for firmware sysfs path construction instead of `fmt::format` string.
- Consolidate `fw_file` and `fw_value` declarations to the top of `read_fw_info`.
- Use `fw_path.string()` in the error message `fmt::format` call.
- Normalize `catch(const std::exception& e)` style.
- Rename `fw_status` to `status` in `ioctl_pcs_create`.
- Move the `ROCPROFILER_STATUS_STRING` macro call to a single line per existing convention.

These are cosmetic but the PR is in `changes_requested` state. Accept or reject them explicitly to unblock re-review.

---

#### 10. `ioctl_query_pcs_configs` firmware check ordering — vlaindic's suggestion thread unresolved

**File:** `source/lib/rocprofiler-sdk/pc_sampling/ioctl/ioctl_adapter.cpp:583`

vlaindic suggested placing the firmware check last in `ioctl_pcs_create` (after KFD version and IOCTL config checks) to keep the logical ordering consistent. The thread is still open.

---

#### 11. `queue_controller.cpp` default agent — vlaindic's "INVALID vs zero" question unresolved

**File:** `source/lib/rocprofiler-sdk/hsa/queue_controller.cpp:146`

vlaindic asked whether the firmware fields on the `default_agent` placeholder should use the INVALID sentinel or zero. The thread has no reply from the author. Acknowledge and close it.

sauverma93 commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/tests/agent.cpp:
```diff
@@ -344,8 +390,76 @@ get_gpu_agents()
+{
+    namespace sdk = ::rocprofiler::sdk;
+
+    auto fw_path = fmt::format("/sys/class/drm/renderD{}/device/fw_version/{}", drm_render_minor, fname);
```
@vlaindic  we use this method to read firmware info,. This path doesn't require sudo access, but downside is we only read firmware version, and don't get feature version

sauverma93 committed:
Add GPU firmware info to agent API and gate PC sampling on MI300

1. Add rocprofiler_agent_firmware_info_v0_t with per-engine firmware version
   fields (MEC, MEC2, RLC, SDMA, SMC, SOS, TA, VCN, etc.)
2. Extend rocprofiler_agent_v0_t with fw_info (GPU-only, appended at end of
   struct; sizeof(rocprofiler_agent_t) 312 -> 368)
3. Add ROCPROFILER_FIRMWARE_VERSION_NONE sentinel for missing firmware data
4. Read firmware versions from sysfs on Linux
   (/sys/class/drm/renderD{N}/device/fw_version/*) during agent enumeration
5. Initialize fw_info to defaults for CPU agents and the HSA queue-controller
   stub agent
6. Add ROCPROFILER_STATUS_ERROR_INCOMPATIBLE_FIRMWARE and human-readable
   status string
7. Gate PC sampling on gfx942 (MI300) by firmware revision:
   - Stochastic: MEC >= 0x000000b9
   - Host-trap: SOS >= 0x00360259
8. Filter incompatible configs from ioctl_query_pcs_configs and reject
   incompatible sessions in ioctl_pcs_create
9. Serialize/deserialize fw_info via cereal load/save helpers
10. Register new status code in enum_string.hpp
11. Add ABI tests for rocprofiler_agent_t and
    rocprofiler_agent_firmware_info_v0_t layout
12. Add agent_fw_info integration test cross-validating populated fields
    against sysfs on GPU agents
13. Add basic fw_info sanity checks in the existing agent test
14. Update PC sampling docs with minimum firmware requirements, rocm-smi
    --showfwinfo usage, and optional debugfs feature-version guidance

vlaindic reviewed (approved):
Thanks a lot for addressing my comments. The overall PR LGTM, so I'm approving. Could nit-comments, but nothing blocking. 

I recall @jrmadsen proposed using sentinel of UINT32_MAX for the FW version, but I believe VERSION_NONE=0 works as well, as you implemented that checks that confirms if the version is missing inside the ioctl_adapter.cpp.

sauverma93 committed:
formatting fix

vlaindic reviewed (commented):
I've already approved, thanks for addressing comments and providing more information about the FW version. 

I believe it would be useful to document this information about FW vs feature version in the code as a future reference.
